File: windows/Hinge.Core/RemoteInputModel.cs
using System.Buffers.Binary;
using System.Text;

namespace Hinge.Core;

public enum RemoteActionType : byte
{
    MouseMove        = 0x01,
    MouseDown        = 0x02,
    MouseUp          = 0x03,
    MouseClick       = 0x04,
    MouseDoubleClick = 0x05,
    MouseScroll      = 0x06,
    KeyDown          = 0x10,
    KeyUp            = 0x11,
    TextInput        = 0x12
}

public enum RemoteMouseButton : byte
{
    Left   = 0,
    Right  = 1,
    Middle = 2
}

public class RemoteInputEvent
{
    public const int HeaderSize = 16;

    public RemoteActionType ActionType { get; set; }
    public byte ButtonOrKey { get; set; }
    public ushort Flags { get; set; }
    public int DeltaX { get; set; }
    public int DeltaY { get; set; }
    public int WheelOrData { get; set; }
    public string? TextPayload { get; set; }

    public byte[] Serialize()
    {
        byte[]? textBytes = null;
        int totalSize = HeaderSize;
        if (ActionType == RemoteActionType.TextInput && !string.IsNullOrEmpty(TextPayload))
        {
            textBytes = Encoding.UTF8.GetBytes(TextPayload);
            totalSize += textBytes.Length;
        }

        byte[] buffer = new byte[totalSize];
        var span = buffer.AsSpan();

        span[0] = (byte)ActionType;
        span[1] = ButtonOrKey;
        BinaryPrimitives.WriteUInt16BigEndian(span.Slice(2, 2), Flags);
        BinaryPrimitives.WriteInt32BigEndian(span.Slice(4, 4), DeltaX);
        BinaryPrimitives.WriteInt32BigEndian(span.Slice(8, 4), DeltaY);
        BinaryPrimitives.WriteInt32BigEndian(span.Slice(12, 4), WheelOrData);

        if (textBytes != null && textBytes.Length > 0)
        {
            textBytes.CopyTo(span.Slice(HeaderSize));
        }

        return buffer;
    }

    public static bool TryParse(ReadOnlySpan<byte> span, out RemoteInputEvent? result)
    {
        result = null;
        if (span.Length < HeaderSize) return false;

        var ev = new RemoteInputEvent
        {
            ActionType = (RemoteActionType)span[0],
            ButtonOrKey = span[1],
            Flags = BinaryPrimitives.ReadUInt16BigEndian(span.Slice(2, 2)),
            DeltaX = BinaryPrimitives.ReadInt32BigEndian(span.Slice(4, 4)),
            DeltaY = BinaryPrimitives.ReadInt32BigEndian(span.Slice(8, 4)),
            WheelOrData = BinaryPrimitives.ReadInt32BigEndian(span.Slice(12, 4))
        };

        if (ev.ActionType == RemoteActionType.TextInput && span.Length > HeaderSize)
        {
            ev.TextPayload = Encoding.UTF8.GetString(span.Slice(HeaderSize));
        }

        result = ev;
        return true;
    }
}

File: android/lib/core/remote_input_model.dart
import 'dart:convert';
import 'dart:typed_data';

enum RemoteActionType {
  mouseMove(0x01),
  mouseDown(0x02),
  mouseUp(0x03),
  mouseClick(0x04),
  mouseDoubleClick(0x05),
  mouseScroll(0x06),
  keyDown(0x10),
  keyUp(0x11),
  textInput(0x12),
  unknown(0x00);

  final int value;
  const RemoteActionType(this.value);

  static RemoteActionType fromValue(int val) {
    return RemoteActionType.values.firstWhere(
      (e) => e.value == val,
      orElse: () => RemoteActionType.unknown,
    );
  }
}

enum RemoteMouseButton {
  left(0),
  right(1),
  middle(2);

  final int value;
  const RemoteMouseButton(this.value);
}

class RemoteInputEvent {
  static const int headerSize = 16;

  final RemoteActionType actionType;
  final int buttonOrKey;
  final int flags;
  final int deltaX;
  final int deltaY;
  final int wheelOrData;
  final String? textPayload;

  RemoteInputEvent({
    required this.actionType,
    this.buttonOrKey = 0,
    this.flags = 0,
    this.deltaX = 0,
    this.deltaY = 0,
    this.wheelOrData = 0,
    this.textPayload,
  });

  Uint8List serialize() {
    Uint8List? textBytes;
    int totalSize = headerSize;
    if (actionType == RemoteActionType.textInput &&
        textPayload != null &&
        textPayload!.isNotEmpty) {
      textBytes = Uint8List.fromList(utf8.encode(textPayload!));
      totalSize += textBytes.length;
    }

    final data = ByteData(totalSize);
    data.setUint8(0, actionType.value);
    data.setUint8(1, buttonOrKey);
    data.setUint16(2, flags, Endian.big);
    data.setInt32(4, deltaX, Endian.big);
    data.setInt32(8, deltaY, Endian.big);
    data.setInt32(12, wheelOrData, Endian.big);

    final bytes = data.buffer.asUint8List();
    if (textBytes != null) {
      bytes.setRange(headerSize, totalSize, textBytes);
    }
    return bytes;
  }

  static RemoteInputEvent? tryParse(Uint8List bytes) {
    if (bytes.length < headerSize) return null;

    final data = ByteData.sublistView(bytes);
    final action = RemoteActionType.fromValue(data.getUint8(0));
    final button = data.getUint8(1);
    final flags = data.getUint16(2, Endian.big);
    final dx = data.getInt32(4, Endian.big);
    final dy = data.getInt32(8, Endian.big);
    final wheel = data.getInt32(12, Endian.big);

    String? text;
    if (action == RemoteActionType.textInput && bytes.length > headerSize) {
      text = utf8.decode(bytes.sublist(headerSize));
    }

    return RemoteInputEvent(
      actionType: action,
      buttonOrKey: button,
      flags: flags,
      deltaX: dx,
      deltaY: dy,
      wheelOrData: wheel,
      textPayload: text,
    );
  }
}
